diff --git a/src/plugins/nat/nat.c b/src/plugins/nat/nat.c
index 2657a70..62a5a7d 100755
--- a/src/plugins/nat/nat.c
+++ b/src/plugins/nat/nat.c
@@ -84,6 +84,7 @@ VNET_FEATURE_INIT (ip4_nat44_ed_classify, static) = {
   .arc_name = "ip4-unicast",
   .node_name = "nat44-ed-classify",
   .runs_after = VNET_FEATURES ("acl-plugin-in-ip4-fa"),
+  .runs_before = VNET_FEATURES ("nat44-ed-in2out", "nat44-ed-out2in"),
 };
 VNET_FEATURE_INIT (ip4_snat_in2out_worker_handoff, static) = {
   .arc_name = "ip4-unicast",
@@ -1749,10 +1750,10 @@ feature_set:
                 feature_name = "nat44-classify";
               }
 
-            vnet_feature_enable_disable ("ip4-unicast", del_feature_name,
-                                         sw_if_index, 0, 0, 0);
             vnet_feature_enable_disable ("ip4-unicast", feature_name,
                                          sw_if_index, 1, 0, 0);
+            vnet_feature_enable_disable ("ip4-unicast", del_feature_name,
+                                         sw_if_index, 0, 0, 0);
             if (!is_inside)
               {
                 if (sm->endpoint_dependent)

